src/main/antora/modules/security/pages/security.adoc

```diff
@@ -1713,4 +1713,223 @@ The test itself is basically the same as that for the <<Securing an endpoint wit
+            </p>
+        </h:form>
+    </h:body>
+</html>
+----
+
+[source,java]
+----
+@Named
+@RequestScoped
+public class LoginBacking {
+
+    @Inject
+    private SecurityContext securityContext;
+
+    @Inject
+    private FacesContext facesContext;
+
+    @NotNull
+    @Size(min = 3, max = 15, message="Username must be between 3 and 15 characters")
+    private String username;
+
+    @NotNull
+    @Size(min = 5, max = 50, message="Password must be between 5 and 50 characters")
+    private String password;
+
+    public void login() {
+        switch (
+            // Continue the authentication dialog manually by invoking the authenticate()
+            // method. The form authentication picks this up, just like a post to j_security does.
+            securityContext.authenticate(
+                getRequest(),
+                getResponse(),
+                withParams()
+                    .credential(new UsernamePasswordCredential(username, new Password(password))))) {
+
+            case SEND_CONTINUE:
+
+                // Authentication mechanism has send a redirect, should not
+                // send anything to response from Faces now.
+                facesContext.responseComplete();
+                return;
+
+            case SEND_FAILURE:
+
+                addError("Login failed");
+                return;
+
+            default:
+        }
+    }
+
+    // getters/setters + utility methods omitted
+}
+----
+
+The view itself is quite similar to the HTML page we used for the form in <<Securing an endpoint with Form authentication>>. The main difference is bindings of the form fields to a (CDI) backing bean. The bean side of the binding has Jakarta Validation constraints applied to it; this allows for fine-grained validation of some general requirements of the credentials without actually attempting authentication.
+
+If this initial validation passes, we arrive in the `login()` method. For our example the only thing we need to do here is signaling that we want to continue the authentication dialog (the process or interaction between the caller and the authentication mechanism), and when doing so provide the credentials that we earlier obtained in a custom way.
+
+We have two important outcomes to handle. `SEND_CONTINUE` effectively means the credentials were validated successfully, and the caller is therefore directed to the resource that was originally requested. `SEND_FAILURE` means the opposite; the credentials were not validated successfully. By just returning from our callback method in the last case the form will be redisplayed, albeit with the error message added.
+
+NOTE: `NOT_DONE` and `SUCCESS` are two other outcomes, but we don't have to handle them here. `NOT_DONE` only applies to pre-emptive authentication, but we're doing explicit (forced, mandatory) authentication here. `SUCCESS` means we can go ahead and render the page, which is exactly what happens if we don't do anything.
+
```

Comment:
Perhaps we want to add a link to 'Getting Started with Web Applications' in case the reader wanted to know how exactly to run this xhtml+bean.